core/src/main/scala/kafka/admin/AdminClient.scala

```diff
@@ -115,6 +124,56 @@ class AdminClient(val time: Time,
     listAllGroupsFlattened.filter(_.protocolType == ConsumerProtocol.PROTOCOL_TYPE)
   }
 
+  class BrokerVersionInfo(val apiVersions: Map[ApiKeys, ApiVersion]) {
+    override def toString(): String = {
+      val bld = new StringBuilder()
+      var prefix = "\n  "
+      bld.append("{");
+      ApiKeys.values().foreach { apiKey =>
+        bld.append(prefix).append(apiKey.name).append("(").
+          append(apiKey.id).append("): ")
+        val apiVersion = apiVersions.get(apiKey)
+        apiVersion match {
+          case None => bld.append("(unsupported)")
+          case Some(version) => if (version.minVersion == version.maxVersion) {
+            bld.append(version.maxVersion);
+          } else {
+            bld.append(version.minVersion).append(" to ").
+                append(version.maxVersion);
+          }
+        }
+        prefix = ",\n  ";
+      }
+      bld.append("\n}")
+      bld.toString()
+    }
+  }
+
+  object BrokerVersionInfo {
+    def create(apiVersionCollection: util.Collection[ApiVersion]): BrokerVersionInfo = {
+      var apiVersions = scala.collection.mutable.Map[ApiKeys, ApiVersion]()
+      apiVersionCollection.foreach { apiVersion =>
+        apiVersions += (ApiKeys.forId(apiVersion.apiKey) -> apiVersion)
+      }
+      new BrokerVersionInfo(apiVersions.toMap)
+    }
+  }
+
+  def listAllBrokerVersionInfo(): Map[Node, BrokerVersionInfo] = {
```

Comment:
Working locally with this patch, I'm still getting an error trying to test these new commands. Seems the API versions don't work -- any idea what's going wrong?

```
java.lang.IllegalStateException: Attempt to send API Versions request to node 0 which is not ready.
	at org.apache.kafka.clients.NetworkClient.doSend(NetworkClient.java:309)
	at org.apache.kafka.clients.NetworkClient.send(NetworkClient.java:296)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.trySend(ConsumerNetworkClient.java:398)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:255)
	at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:181)
	at kafka.admin.AdminClient.kafka$admin$AdminClient$$send(AdminClient.scala:49)
	at kafka.admin.AdminClient.getApiVersions(AdminClient.scala:83)
	at kafka.admin.AdminClient$$anonfun$listAllBrokerVersionInfo$1.liftedTree2$1(AdminClient.scala:166)
	at kafka.admin.AdminClient$$anonfun$listAllBrokerVersionInfo$1.apply(AdminClient.scala:165)
	at kafka.admin.AdminClient$$anonfun$listAllBrokerVersionInfo$1.apply(AdminClient.scala:162)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at kafka.admin.AdminClient.listAllBrokerVersionInfo(AdminClient.scala:162)
	at kafka.admin.BrokerVersionCommand$.main(BrokerVersionCommand.scala:37)
	at kafka.admin.BrokerVersionCommand.main(BrokerVersionCommand.scala)
```